submissions/nadsatt/pre-oop/index.js

```diff
@@ -0,0 +1,99 @@
+/* 
+   Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: https://github.com/nadsatt/a-tiny-JS-world
+   Web app: https://nadsatt.github.io/a-tiny-JS-world/
+*/
+
+// ======== OBJECTS DEFINITIONS ========
+class Inhabitant {
+   constructor({species, hands = 0, legs = 0, saying, gender, name, friends = []}){
+      this.species = species;
+      this.name = name;
+      this.gender = gender;
+      this.saying = saying;
+      this.hands = hands;
+      this.legs = legs;
+      this.friends = friends;
+   }
+
+   toString(){
+      return Object.entries(this).reduce((res, [key, value]) => {
+
+         if(key === 'friends' && value.length === 0) res += `${key}: no friends; `
+         else if (key === 'friends') res += `${key}: ${value.map(obj => obj.name).join(',')}; `
+         else res += `${key}: ${value}; `
+
+         return res;
+      }, '');
+   }
+}
+
+class Cat extends Inhabitant {
+   constructor(gender, name, friends){
+      super({species: 'cat', legs: 4, saying: 'meov-meov!', gender, name, friends});
+   }
+}
+
+class Dog extends Inhabitant {
+   constructor(gender, name, friends){
+      super({species: 'dog', legs: 4, saying: 'woof-woof!', gender, name, friends});
+   }
+}
+
+class Human extends Inhabitant {
+   constructor({saying, name, gender, friends}){
+      super({species: 'human', hands: 2, legs: 2, saying, gender, name, friends});
+   }
+}
+
+class Man extends Human {
+   constructor(name, friends){
+      super({saying: 'i\'m a man!', gender: 'male', name, friends});
+   }
+}
+
+class Woman extends Human {
+   constructor(name, friends){
+      super({saying: 'i\'m a woman!', gender: 'female', name, friends});
+   }
+}
+
+class CatWoman extends Cat {
+   constructor(name, friends){
+      super('female', name, friends);
+      this.species = 'human';
+      this.hands = 2;
+      this.legs = 2;
+   }
+}
+
+const inhabitants = [
+   new Cat('female', 'Kitty'),
+   new Dog('male', 'Bob', [new Dog('female', 'Stella')]),
+   new Man('Nick', [new Dog('male','Baxter'), new Cat('male', 'Simon')]),
+   new Woman('Lucy'),
+   new CatWoman('Kate', [new Man('Tom')])
+];
+
+// ======== OUTPUT ========
+/* 
+   Use print(message) for output.
+   Default tag for message is <pre>. Use print(message,'div') to change containing element tag.
+
+   Message can contain HTML markup. You may also tweak index.html and/or styles.css.
+   However, please, REFRAIN from improving visuals at least until your code is reviewed
+   so code reviewers might focus on a single file that is index.js.
+*/
+
+inhabitants.forEach(inhabitant => print(inhabitant.toString(), 'div'));
+
+function print(strObj, el = 'pre'){
```

Comment:
Thanks for notes and link to materials! Here my questions to re-check if i understood everything right:

1) I can't understand the first point. When you say 'child classes introduce and operate properties unique to them' do you mean *unique properties* or *unique property values*? Because in my code `Inhabitant` super class holds and operates set of properties shared by each extending class. Each instance of `Human`, `Man`, `Woman`, `Cat`, `Dog` and `CatWoman` class contains same 8 properties. There is no unique properties introduced in extending classes (only predefined values assigned by classes). So i can't understand what's wrong with my classes in terms of classes concept. 

2) Should i *(in terms of next oop task)* REPLACE defining `friends` property through the constructor WITH defining `friends` property through the prototype method? 

3) Should i *(in terms of next oop task)* not to pass objects or links to them in constructors at all? 

4) Should i *(in terms of next oop task)* extend superclass' `toString` method using this logic?  
![method-overriding](https://user-images.githubusercontent.com/53071575/98596036-876c3880-22df-11eb-9516-b5a8c3e9eecd.png)

Reply:
1. **Unique properties**
Hands is a property uniquely pertaining to humans. Thus they should have been declared for humans only. We cannot ignore the domain setting. The beauty of OOP approach is that, in particular, each class has what it it needs only, inheriting shared properties from a super class (or a chain of super classes) and introducing what is unique to a given class.

Reply:
I got it, thanks!

Reply:
2. **Friends**
The approach to the addition of friends is pretty OK.
A minor issue is that there are more inhabitants created in your code than exposed to the reader (e.g. Tom is created but not exposed as the world's inhabitant, only a (distant?) friend to someone.
Constructor is not the only method to populate object's properties. You might a method to add friends at a later time. This way you can make your inhabitants cross-friends.

Reply:
3. **Passing object**
Passing an object like a friend is OK.
I meant replacing constructions like `constructor({species, hands = 0, legs = 0, saying, gender, name, friends = []})` with something like `constructor(species, hands = 0, legs = 0, saying, gender, name, friends = [])`, i.e. properties not packed in a properties object. You'd need to regroup arguments order so that properties that **may** have default values come after mandatory arguments.

In general this is quite legit, but for the purposes of methodology this may make some thing entangled when explaining OOP fundamentals and identifiying source of issues.

Reply:
4. **`toString`**
In short, yes. Not necessarily as illustrated but something like following would work.

```
class Human extends Inhabitant {
  contructor(name, gender, saying) {
    // .....
  }
  toString() {
    return super.toString() + this.hands + this.legs;
  }
}
```

Please note the `toString` method code above is much simplified for the sake of focusing on major concepts.

Reply:
Please let me know if you have any further questions.
You do not need to fix anything of the above in this task. You will take your effort on this later.

Reply:
Okay, now i understood everything. One more thanks for clarifying.